dbms/src/DataStreams/ScanHashMapAfterProbBlockInputStream.cpp

```diff
@@ -317,22 +395,31 @@ void NonJoinedBlockInputStream::fillColumns(const Map & map,
 
     for (; *it != end;)
     {
-        if ((*it)->getMapped().getUsed())
+        if constexpr (row_flagged && output_joined_rows)
+            row_count_info.inc(AdderMapEntryRightSemi<true, typename Map::mapped_type>::add((*it)->getMapped(), key_num, num_columns_left, mutable_columns_left, num_columns_right, mutable_columns_right, next_element_in_row_list, row_count_info.availableRowCount()));
+        else if constexpr (row_flagged && !output_joined_rows)
+            row_count_info.inc(AdderMapEntryRightSemi<false, typename Map::mapped_type>::add((*it)->getMapped(), key_num, num_columns_left, mutable_columns_left, num_columns_right, mutable_columns_right, next_element_in_row_list, row_count_info.availableRowCount()));
+        else
         {
-            ++(*it);
-            continue;
+            bool mapped = (*it)->getMapped().getUsed();
+            if ((mapped && !output_joined_rows) || (!mapped && output_joined_rows))
+            {
+                ++(*it);
+                continue;
+            }
+
+            row_count_info.inc(AdderMapEntry<STRICTNESS, typename Map::mapped_type>::add(
+                (*it)->getMapped(),
+                key_num,
+                num_columns_left,
+                mutable_columns_left,
+                num_columns_right,
+                mutable_columns_right,
+                next_element_in_row_list,
+                row_count_info.availableRowCount()));
+            assert(row_count_info.getCurrentRows() <= max_block_size);
```

Comment:
This assert does not work for `AdderMapEntryRightSemi `?

Reply:
It works, moved it to a common code path